### PR TITLE
fix: plainify error objects for console

### PR DESCRIPTION
**Pull request by Shou**
We make error objects into plain objects and `JSON.stringify` them the
same, such that they no longer appear as empty objects.

### CODE REVIEW

File: tests/utils.js [hunk 1]
```diff
@@ -157,6 +157,16 @@ describe('Utils', () => {
...
+    it('should return a plain object when given an error object', function() {
+      const error = new Error('Hello, World!');
+      const hiddenPaths = utils.hideAbsolutePathsInObject(error);
+
+      chai.expect(hiddenPaths).to.deep.equal({
+        message: 'Hello, World!',
+        stack: error.stack
+      });
+    });
+
```
Comment (lurch) [inline tests/utils.js:169]: Perhaps it'd also be worth testing an error containing an absolute path?

File: src/utils.js [hunk 2]
```diff
@@ -115,6 +115,18 @@ exports.flattenStartCase = (object) => {
  * > }
  */
 exports.hideAbsolutePathsInObject = (object) => {
+  if (_.isError(object)) {
+
+    // Turn the Error into an Object
+    const error = {};
+
+    _.map(Object.getOwnPropertyNames(object), (key) => {
```
Comment (jviotti) [inline src/utils.js]: Using a map without assigning its result looks weird. Let's either use `_.each`, or convert this into a fold:

```js
const error = _.reduce(Object.getOwnPropertyNames(object), (accumulator, key) => {
    accumulator[key] = object[key]
    return result;
}, {})
```

Comment (Shou) [reply src/utils.js]: Using folds is great here!

File: tests/utils.js (hunk 1 above)
Comment (Shou) [reply tests/utils.js:169]: Very good idea, I had a sneaking suspicion about paths in error stack traces, and this confirms it: it can't handle them as-is. I think we'll have to treat error stack traces as an edge-case because there are several paths riddled throughout and they start with characters that are valid in paths but aren't actually part of the path, namely the opening-parens.

```
Error: Hello, World!
    at Context.<anonymous> (/home/shou/Prog/Javascript/resin-corvus/tests/utils.js:171:21)
    at callFn (/home/shou/Prog/Javascript/resin-corvus/node_modules/mocha/lib/runnable.js:345:21)
    at Test.Runnable.run (/home/shou/Prog/Javascript/resin-corvus/node_modules/mocha/lib/runnable.js:337:7)
    at Runner.runTest (/home/shou/Prog/Javascript/resin-corvus/node_modules/mocha/lib/runner.js:444:10)
    at /home/shou/Prog/Javascript/resin-corvus/node_modules/mocha/lib/runner.js:550:12
    at next (/home/shou/Prog/Javascript/resin-corvus/node_modules/mocha/lib/runner.js:361:14)
    at /home/shou/Prog/Javascript/resin-corvus/node_modules/mocha/lib/runner.js:371:7
    at next (/home/shou/Prog/Javascript/resin-corvus/node_modules/mocha/lib/runner.js:295:14)
    at Immediate.<anonymous> (/home/shou/Prog/Javascript/resin-corvus/node_modules/mocha/lib/runner.js:339:5)
    at runCallback (timers.js:651:20)
    at tryOnImmediate (timers.js:624:5)
    at processImmediate [as _immediateCallback] (timers.js:596:5)
```

A naive way I see of solving it is to split on opening-parentheses and treat what's in-between as a potential path, and `basename` it. The rest can be done on a per-word basis, presumably. 

Edit
Spaces in paths complicate matters. It truncates the path to the first space, but from there the path is kept intact.

**Comment by Shou** [comment]
Here's what I have so far though. To reiterate, it doesn't handle paths with spaces in them.

File: src/utils.js [hunk 3]
```diff
@@ -85,6 +85,41 @@ exports.flattenStartCase = (object) => {
...
+ * @function
+ * @public
+ *
+ * @param {String} str - string potentially containing paths
+ * @returns {String} string with basenamed paths if any
+ *
+ * @example
+ */
+exports.removeContainedPaths = (str) => {
+  const absolutify = (pathname) => {
```
Comment (lurch) [inline src/utils.js]: `absolutify` is the wrong name here isn't it? Should it be `relativify` or `make_relative`?

File: src/utils.js [hunk 4]
```diff
@@ -85,6 +85,41 @@ exports.flattenStartCase = (object) => {
...
+ * @summary Remove paths contained in strings
+ * @function
+ * @public
+ *
+ * @param {String} str - string potentially containing paths
+ * @returns {String} string with basenamed paths if any
+ *
+ * @example
+ */
+exports.removeContainedPaths = (str) => {
```
Comment (lurch) [inline src/utils.js]: And I guess this is `relativifyContainedPaths` rather than `removeContainedPaths`?

EDIT: Or perhaps `hideAbsolutePathsInString` ?

File: src/utils.js (hunk 3 above)
Comment (Shou) [reply src/utils.js]: You're right, I wanted to push out the changes without brushing them up because I hit the paths-with-spaces issue and wanted feedback on that. I'll make sure to change it.

File: src/utils.js [hunk 5]
```diff
@@ -85,6 +85,41 @@ exports.flattenStartCase = (object) => {
 };
 
 /**
+ * @summary Remove paths contained in strings
```
Comment (lurch) [inline src/utils.js]: nit: Need to update this `@summary`

File: tests/utils.js [hunk 6]
```diff
@@ -157,6 +157,40 @@ describe('Utils', () => {
       chai.expect(utils.hideAbsolutePathsInObject(object)).to.deep.equal(object);
     });
 
+    it('should return a plain object when given an error object', function() {
+      const error = new Error('Hello, World!');
+      const hiddenPaths = utils.hideAbsolutePathsInObject(error);
+
+      chai.expect(hiddenPaths).to.deep.equal({
+        message: 'Hello, World!',
+        stack: utils.removeContainedPaths(error.stack)
```
Comment (lurch) [inline tests/utils.js]: Yuck, would it be possible to do this without making function calls inside the `expect to equal` section?

File: src/utils.js [hunk 7]
```diff
@@ -85,6 +85,41 @@ exports.flattenStartCase = (object) => {
...
+ * @function
+ * @public
```
Comment (lurch) [inline src/utils.js]: Does this need to be public, or could it be private?
I think if it needs to be public, then it ought to have it's own set of unit-tests? (ideally)

**Comment by lurch** [comment]
Oh! Does this repo not have CI tests enabled?

**Comment by jviotti** [comment]
-.- @Shou Do you mind quickly enabling Travis and Appveyor here?

**Comment by lurch** [comment]
Presumably you'll be rebasing this, so that the CI tests now get run? :)

**Comment by Shou** [comment]
Yup! I'm working on making some other changes here now too.

**Comment by Shou** [comment]
This should be good to go then, since we're not worrying about path anonymization here, unless anyone has any final comments.

**Review by jviotti — APPROVED** [review]

**Comment by lurch** [comment]
So I guess the path-anonymisation stuff will be moved to a separate PR?

**Comment by lurch** [comment]
The 'reduced' version of this PR has been merged - I still think a new PR ought to be created, containing the anonymisation-of-paths-in-errors stuff that this PR _used_ to contain. @Shou 
(hopefully I'm not being overly pedantic)

**Comment by Shou** [comment]
I've pushed the branch, but it's not ready for a PR yet as I'm in the middle of some changes there.
